### PR TITLE
WIP: IO mapping fixes

### PR DESCRIPTION
**Pull request by ret2libc**
Based on https://github.com/radare/radare2/pull/10355/ . Opening a new one here so that every maintainer can work on this branch.

r2r PR: https://github.com/radare/radare2-regressions/pull/1405

### CODE REVIEW

**Comment by ret2libc** [comment]
Yes I know, we are going to squash everything at the end, but right now it's just a waste of time resolving conflicts for every rebased commits, merge is just easier.

**Comment by condret** [comment]
I think this branch is an improvement, should I rebase this?

**Comment by radare** [comment]
and squash and fix the conflicts too. this cant be merged right now

File: libr/core/cmd_open.c [hunk 1]
```diff
@@ -205,15 +221,18 @@ static void list_maps_visual(RIO *io, ut64 seek, ut64 len, int width, int use_co
 
 	// seek = (io->va || io->debug) ? r_io_section_vaddr_to_maddr_try (io, seek) : seek;
 
-	ls_foreach_prev (io->maps, iter, s) {			//this must be prev, maps the previous map allways has lower priority
-		min = R_MIN (min, s->itv.addr);
-		max = R_MAX (max, r_itv_end (s->itv) - 1);
-	}
+	mm.min = min;
+	mm.max = max;
+	r_oids_foreach(io->maps, _map_min_max_cb, &mm);
```
Comment (radare) [inline libr/core/cmd_open.c]: missing space before (

File: libr/core/cmd_open.c [hunk 2]
```diff
@@ -436,16 +456,18 @@ static void cmd_open_bin(RCore *core, const char *input) {
...
 	if (!io || !io->maps || !print || !print->cb_printf) {
 		return;
 	}
 	if (mode == 'j') {
 		print->cb_printf ("[");
 	}
 	bool first = true;
-	ls_foreach_prev (io->maps, iter, map) {			//this must be prev
+	od = io->maps->ptop - 1;
+	while ((map = r_oids_oget(io->maps, od))) {
```
Comment (radare) [inline libr/core/cmd_open.c]: missing space

**Comment by radare** [comment]
please rebase and fix the conflicts

**Comment by radare** [comment]
can we close the other pr?

**Comment by radare** [comment]
rebased

File: libr/io/submap.c [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+static ut64 _map_from(RIOMap *map) {
+	return map->itv.addr;
+}
+
+static ut64 _map_to(RIOMap *map) {
+	return map->itv.addr + map->itv.size - 1;
+}
+
+//in - incoming
+static bool _submap_find_cb(void *vin, void *vincoming, void *user, int *cmp_res) {
```
Comment (MaskRay) [inline libr/io/submap.c]: Having side effects in comparison functions is bad.

File: libr/io/submap.c [hunk 4]
```diff
@@ -0,0 +1,289 @@
+#include <r_io.h>
+#include <r_types.h>
+#include <r_util.h>
+
+static ut64 _map_from(RIOMap *map) {
+	return map->itv.addr;
+}
+
+static ut64 _map_to(RIOMap *map) {
+	return map->itv.addr + map->itv.size - 1;
```
Comment (MaskRay) [inline libr/io/submap.c]: `- 1` is ugly

File: libr/io/submap.c [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+	if (off[0] > sm->to) {
+		cmp_res[0] = -1;
+		goto beach;
+	}
+	cmp_res[0] = 0;
+beach:
+	return true;
+}
+
+static bool _submap_free_cb(void *user, void *data, ut32 id) {
```
Comment (MaskRay) [inline libr/io/submap.c]: I don't see why this function is necessary

File: libr/io/submap.c [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+	io->submaps->cmp = _submap_insert_no_overlap_cb;
+	ut32 od;
+	r_oids_insert (io->submaps, sm, &sm->sid, &od, NULL);
+}
+
+//cuts out all submaps, so that there is no submap between from and to
+R_API bool r_io_submap_cut_out(RIO *io, ut64 from, ut64 to) {
+	RIOSubMap *bd, *sm;
+	ut32 od;
+	if (to < from) {
```
Comment (MaskRay) [inline libr/io/submap.c]: unnecessary check

Comment (MaskRay) [reply libr/io/submap.c]: Handling address space wraparound here is terrible

File: libr/io/submap.c [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+				io->submaps->cmp = _submap_insert_no_overlap_cb;
+				r_oids_insert (io->submaps, bd, &bd->sid, &od, NULL);
+				return true;
+			}
+//    ##cut this out##
+//###sm###
+//becomes
+//    ##cut this out##
+//#sm#
+			sm->to = from - 1;
```
Comment (MaskRay) [inline libr/io/submap.c]: This is incorrect if sm->from < sm->to < from < to

File: libr/io/submap.c [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+//any further submap is going to be bigger than sm,
+//no need to continue the loop
+		}
+		return true;
+	}
+	return true;
+}
+
+//in - incoming
+static bool _submap_hack_n_slash_insert(void *vin, void *vincoming, void *user, int *cmp_res) {
```
Comment (MaskRay) [inline libr/io/submap.c]: Having side effects in comparison functions is bad.

File: libr/io/submap.c [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+#if 0
+	########in#########
+#############incoming###############
+#endif
+			//slash submap into 2 submaps, that don't intersect with the existing submap in
+			map = R_NEW0 (RIOSubMap);	//check needed
+			map->id = incoming->id;
+			map->to = incoming->to;
+			map->from = in->to + 1;
+			r_queue_enqueue (todo, map);	//will be inserted on the later iterations
```
Comment (MaskRay) [inline libr/io/submap.c]: The queue used here makes reasoning about the logic difficult. The existing skyline is simpler.

**Comment by MaskRay** [comment]
`include/r_util/`r_idpool.h` (data array + permutation array + other stuff) adds an unnecessary abstraction layer which does not decrease time complexity at all (`oids_od_binsert`), but brings much complexity. A plain sorted array (the current representation of IO maps) is simpler than this one.

**Comment by radare** [comment]
Could you explain/document skyline? I also find it confusing. But what worries me is that current io is not correct, so it needs fixing. How do you suggest to fix the actual issue?

**Comment by radare** [comment]
Agree. Theres an ritv api already and we should use that instead

**Comment by ret2libc** [comment]
@maskray please could you write some doc explaining skyline?

@radare what are the issues you are referring to? Do we have actual broken tests in r2r? I created one broken test some time ago in the io unit , but I think we should document all the others to understand why/how it’s necessary to fix things

**Comment by MaskRay** [comment]
Ｉthink I have explained many times but the records are lost (or you ask me to explain it again and again).

Maps have priorities and can overlap, thus to be efficient we should extract the un-shadowed parts, which is called skyline.

**Comment by radare** [comment]
One of the bugs is this:
```
[0x00000000]> cat ./test-io-bug2.r2
o malloc://3
b 3
wow 90
p8
o malloc://1 1 rw
p8
omp 1
p8

CURRENT OUTPUT: 
3
909090
4
900090
900090

expected output:
3
909090
4
900090
909090
```

**Comment by MaskRay** [comment]
> o malloc://3

https://github.com/radare/radare2/pull/11230 will fix this.

I think r_oids* *submaps* in the PR are overengineering and don't improve the IO situation.

I am mostly satisfied with the current high-level design of IO, but am uncomfortable for a few parts:

* `r_io_al_*` `RIOAccessLog` there were introduced nearly 1 year ago, and it turns out that nobody uses the API. I suggest removing them.
* Redundant `r_io_map_cleanup` calls in `r_io_vwrite_at` `r_io_vread_at_mapped`. They harm performance and should be removed later.
* address space wraparound. I guess you all consider `on_map_skyline` complicated. Well, some of the complexity comes from the support of address space wraparound. It can actually be simplified a lot by moving the wraparound part to its callers.

**Comment by ret2libc** [comment]
Honestly... I would close this PR for now. It does too many things in one shot (add new data structures, reimplement skyline, rename stuff, introduce IDs, etc.) and it is too hard to review, nor I do see reasons to change so much the current behaviour.

I think changes should be done in small PRs. I mean, is this PR supposed to fix stuff? How can we really know it won't only hurt more?

**Comment by condret** [comment]
gross

**Comment by condret** [comment]
the current design of rebuilding everything over and ever again is bad, iterating over lists instead of getting stuff in O(log(n)) from a sorted structure is bad for getting maps at a certain address. not having ids and exposing pointers to maps outside of io most of the time is bad.

**Comment by condret** [comment]
and oids is not a datastructure introduced in this pr. permuations are simple.

**Comment by ret2libc** [comment]
@condret oids are a data structure that was never used/tested anywhere. So it's like it was introduced in this PR afaik.

Moreover, just do PRs for each of the things you mentioned and not all together.

**Comment by condret** [comment]
no, some things just only work when they are shipped together

**Pull request closed without merging**